Add a post about changes to WebAssembly targets (#1385)
rust-lang/blog.rust-lang.org: posts/2024-08-26-webassembly-targets-and-new-on-by-default-features.md

---
layout: post

> title: "WebAssembly targets and new on-by-default features"

Reviewer: Discussion: should this use wasm terminology as in "proposals" or "WebAssembly proposals"? Or is target features less confusing?

Reviewer 2: Maybe we should use the common terminology of "target features" so that it's more approachable.

"WebAssembly targets: change in default target-features"

Reviewer: Yeah, I don't feel too strongly about which one, but I think we should stick to just one of them and not mix them.

Author: I've retitled the post, tried to be more consistent internally, and added a bit of discussion on proposal-vs-feature, but let me know if anything else needs to be improved!

> author: Alex Crichton

Reviewer 2: We should mention that you're writing this blog post on the behalf of the compiler team 
```suggestion
author: Alex Crichton
team: The Compiler Team <https://www.rust-lang.org/governance/teams/compiler>
```

---

The Rust compiler has [recently upgraded to using LLVM 19][llvm19] and this
change accompanies some updates to WebAssembly targets of the Rust compiler.
Nightly Rust, what will be come Rust 1.82 on 2024-10-17, reflects all of these
changes and can be used for testing.

WebAssembly is an evolving standard where new features are being added over time
through a [proposals process][proposals]. As WebAssembly proposals reach
maturity, get merged into the specification itself, get implemented in engines,
and remains this way for quite some time then producer toolchains (e.g. LLVM)

> are going to update to include these new proposals by default. In LLVM 19 this

Reviewer: Suggestion: emphasis

```suggestion
are going to update to **enable these sufficiently-mature proposals by default**. In LLVM 19 this
```

has happened with the [multi-value and reference-types proposals][llvmenable].
These are now enabled by default in LLVM and transitively means that it's
enabled by default for Rust as well.

WebAssembly targets for Rust now [have improved
documentation](https://github.com/rust-lang/rust/pull/128511) about WebAssembly
features and disabling them, and this post is going to review these changes and
go into depth about what's changing in LLVM.

## Enabling Reference Types by Default

The [reference-types proposal to
WebAssembly](https://github.com/webAssembly/reference-types) introduced a few
new concepts to WebAssembly, notably the `externref` type which is a
host-defined GC resource that WebAssembly cannot access but can pass around.
Rust does not have support for the WebAssembly `externref` type and LLVM 19 does
not change that. WebAssembly modules produced from Rust will continue to not use

> the `externref` type nor have a means of being able to do so.

Reviewer: Question: is this going to *always* be the case, or for the foreseeable future?

Author: I'll clarify that there are no plans for this but even if it were done it likely wouldn't affect modules by default.

was encoded with a fixed zero byte in its instruction (required to be exactly
0x00). This fixed zero byte was relaxed to a 32-bit [LEB] to indicate which
table the `call_indirect` instruction was using. For those unfamiliar [LEB] is a
way of encoding multi-byte integers in a smaller number of bytes for smaller
integers. For example the integer 0 can be encoded as `0x00` with a [LEB].
[LEB]s are flexible to additionally allow "overlong" encodings so the integer 0
can additionally be encoded as `0x80 0x00`.

LLVM's support of separate compilation of source code to a WebAssembly binary
means that when an object file is emitted it does not know the final index of
the table that is going to be used in the final binary. Before reference-types
there was only one option, table 0, so `0x00` was always used when encoding
`call_indirect` instructions. After reference-types, however, LLVM will emit an
over-long [LEB] of the form `0x80 0x80 0x80 0x80 0x00` which is the maximal
length of a 32-bit [LEB]. This [LEB] is then filled in by the linker with a
relocation to the actual table index that is used by the final module.

> When putting all of this together it means that LLVM 19, which has
> reference-types enabled by default, then any WebAssembly module with an indirect

Reviewer: Suggestion: this read a bit awkward

```suggestion
When putting all of this together, it means that with LLVM 19, which has
reference-types enabled by default, any WebAssembly module with an indirect
```

tool such as [`wasm-opt`] to shrink the size of the output of LLVM. Even before
this change with reference-types it's recommended to do this as [`wasm-opt`] can
typically optimize LLVM's default output even further. When optimizing a module
through [`wasm-opt`] these 5-byte encodings of index 0 are all shrunk to a
single byte.

## Enabling Multi-Value by Default

The second feature enabled by default in LLVM 19 is multi-value. The
[multi-value proposal to WebAssembly][multi-value] enables functions to have
more than one return value for example. WebAssembly instructions are
additionally allowed to have more than one return value as well. This proposal
is one of the first to get merged into the WebAssembly specification after the
original MVP and has been implemented in many engines for quite some time.

The consequences of enabling this feature by default in LLVM are more minor for
Rust, however, than enabling reference-types by default. LLVM's default ABI for
WebAssembly code is not changing even when multi-value is enabled. Additionally
Rust's ABI is not changing either and continues to match LLVM's. Despite this

> though the change has the possibility of still affecting Nightly users of Rust.

Reviewer: Suggestion: I think we want to be careful with implied ABI stability:

- It happens that LLVM's WebAssembly ABI was not changed after multi-value proposal was enabled.
- It happens that Rust's WebAssembly ABI did not change either. (But I think we explicitly don't promise ABI stability for `repr(Rust)`)

*But* we probably want a remark about the stability of Rust's WebAssembly ABIs being subject to change, as I can't imagine this can be perma-frozen with new proposals being added and enabled-by-default?

Author: I've added a small extra section about this to help clarify in addition to trying to explicitly talk about `extern "C"` is the thing not changing. (but please take a look once I push to review as well)

This is not the first time that a WebAssembly proposal has gone from
off-by-default to on-by-default in LLVM, nor will it be the last. For example
LLVM already enables the [sign-extension proposal][sign-ext] by default which
MVP WebAssembly did not have. It's expected that in the not-too-distant future
the
[nontrapping-fp-to-int](https://github.com/WebAssembly/nontrapping-float-to-int-conversions)
proposal will likely be enabled by default. These changes are currently not made
with strict criteria in mind (e.g. N engines must have this implemented for M
years), and there may be breakage that happens.

If you're using a WebAssembly engine that does not support the modules emitted
by Nightly Rust and LLVM 19 then your options are:

* Try seeing if the engine you're using has any updates available to it. You
  might be using an older version which didn't support a feature but a newer
  version supports the feature.
* Open an issue to raise awareness that a change is causing breakage. This could
  either be done on your engine's repository, the Rust repository, or the
  WebAssembly
  [tool-conventions](https://github.com/WebAssembly/tool-conventions)

>   repository.

Reviewer: Suggestion: ... if there isn't an issue about the change/breakage already.
